### DISCUSSION TITLE
Use a callback for an external function call results in 'eval_sx' not defined for CallbackInternal

### DISCUSSION THREAD

**Discussion by nmodry** [casadi/casadi#3943] [Q&A] [answered]
Hello everyone,

I need help getting my callback to work. To set the scene: I am currently working on my master's thesis, for which I need to run some experiments using the [do-mpc](https://www.do-mpc.com/en/latest/) framework. From what I understand, this uses CasADi internally for the computation. One of my controllers is supposed to optimize against the [PMV measure for thermal comfort](https://en.wikipedia.org/wiki/Thermal_comfort#PMV/PPD_method). I wanted to use the [pythermalcomfort](https://github.com/CenterForTheBuiltEnvironment/pythermalcomfort) package to calculate this measure. Sadly, pythermalcomfort works with pure Python code. I've read that the best practice would be to take the pythermalcomfort implementation and rewrite it to work with CasADi (and I've attempted to do so for several hours), but this seems to be way over my head, and I don't have more time to do this. So, my plan B was to use a Callback to make the purely numerical calculation. I will put my code down below. I followed the [documentation example](https://web.casadi.org/docs/#sec-callback) and ran into this error:
`RuntimeError: Error in Function::call for 'pmv_function' [CallbackInternal] at .../casadi/core/function.cpp:370:
.../casadi/core/function_internal.cpp:2024: 'eval_sx' not defined for CallbackInternal`
The FAQ article that led me to the whole callback idea mentioned that this might happen and referred to [this other FAQ post](https://github.com/casadi/casadi/wiki/FAQ:-how-to-resolve-'eval_sx'-not-defined%3F) that I, unfortunately, do not understand.

So my question is: Can you guys help me make a callback that takes two CasADi SX values (dry bulb air temperature and radiant temperature), performs a purely numerical evaluation using both, and then outputs one single CasADi value representing the PMV value, which I then want to use in the controller's optimization function?

Here are the code snippets:

The callback:
```
class PmvCallback(Callback):
    def __init__(self, name, opts={}):
        Callback.__init__(self)
        self.construct(name, opts)

    def get_n_in(self): return 2  # receive to inputs, dry bulb air temp and radiant temp

    def get_n_out(self): return 1  # output just the pmv

    def eval(self, arg):
        T_i, T_m = arg[0], arg[1]
        pmv, _ = utils.ThermalComfort.calc_pmv_ppd(T_i, T_m)
        return [pmv]
```

The code where it's used:
```
T_i = model.x['T_i']
T_m = model.x['T_m']

pmv_function = PmvCallback("pmv_function")
pmv = pmv_function.call([T_i, T_m])

lterm = costs + pmv

mpc.set_objective(mterm=DM(0), lterm=lterm)
```

`T_i` and `T_m` are CasADi SX types.

I've never worked with CasADi or a CAS system before, so I appreciate "complete-newbie" answers. I can also provide more code, answers and even the rewrite I attempted.

**Comment by jgillis** [comment, marked as answer]
Not a do-mpc user, but I think you want to construct you Model with keyword `symvar_type` set to "MX".

Coincidentally, I was just implementing #1869 so that your use-case may actually work out of the box soon.

**Reply by nmodry** [reply]
Thanks; that was the solution. I'm kind of embarrassed that I didn't spot this constructor parameter myself. :smile: 

I am facing a different issue now, but I'll open a new thread for this.
